### PR TITLE
fix(gatsby-dev-cli): refactor, wait for package installation before copying files, handle packages that are not direct deps of local project

### PR DESCRIPTION
**Pull request by pieh**
There's lot of things here:
 - some bug fixes - like waiting for package installation to finish before copying files directly
 - refactoring - move some utilities to separate files (easier to read code and also to test/mock)
 - finally, proper handling of packages that are not direct dependencies of current project (i.e. `gatsby-dev --packages gatsby-cli` will correctly watch `gatsby-cli`, but install `gatsby`)
 - also, clean up .js/.js.map files after installing packages from node_modules for packages that will copy files over to get rid of any stale files (this solves issue @stefanprobst hit in https://github.com/gatsbyjs/gatsby/pull/14283#issuecomment-495398172 )

Hopefully added test coverage add some clarity and confidence

/cc @wardpeet

### CODE REVIEW

File: packages/gatsby-dev-cli/src/__tests__/watch.js [hunk 1]
```diff
@@ -142,11 +172,487 @@ describe(`watching`, () => {
...
+    installPackages.mockClear()
+    checkDepsChanges.mockClear()
+    promisifiedSpawn.mockClear()
+  })
+
+  afterAll(() => {
+    global.process = realProcess
+  })
+
+  describe(`publishing and installing packages`, () => {
```
Comment (wardpeet) [inline packages/gatsby-dev-cli/src/__tests__/watch.js]: should we also check if gatsby copy is working?

Comment (pieh) [reply packages/gatsby-dev-cli/src/__tests__/watch.js]: Like checking if `fs.copy` was called from monorepo to local project? If so I think we can do that in `installs from npm before copying files` test (last test from this file) - it also sounds like good idea

Comment (wardpeet) [reply packages/gatsby-dev-cli/src/__tests__/watch.js]: yes check if copy has been for the correct packages.

so for example we should test
```
checkDepsChanges.mockImplementationOnce(mockDepsChanges([`gatsby`]))
watch(process.cwd(), [`gatsby`, `gatsby-plugin-sharp`], {
  scanOnce: true,
  quiet: true,
  monoRepoPackages,
  localPackages: [`gatsby`, `gatsby-plugin-sharp`],
})
```

now it should install gastby and copy gatsby & gatsby-plugin-sharp

Comment (pieh) [reply packages/gatsby-dev-cli/src/__tests__/watch.js]: gotcha! makes total sense, will add

Comment (pieh) [reply packages/gatsby-dev-cli/src/__tests__/watch.js]: added https://github.com/gatsbyjs/gatsby/pull/14347/commits/36542cfa0b20b719e58f99ca793f0545985b2ced - let me know if this is what you had in mind

**Comment by m-allanson** [comment]
@stefanprobst this wasn't published earlier, now available in `gatsby-dev-cli@2.4.16`

**Pull request merged**